### PR TITLE
Fixed rabbitmq upgrade

### PR DESCRIPTION
**Pull request by atsikham**
Tasks #1979 #1984. Made it in the one PR as fixes in the same place are necessary.
Tried to find universal way to check clustering and rabbitmq version.

### CODE REVIEW

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml]: How about parsing json payload? :thinking: 
```shell
root@u1a2:~# rabbitmqctl cluster_status --formatter json
{"alarms":[],"cluster_name":"rabbit@u1a2","disk_nodes":["rabbit@u1a2","rabbit@u1a3","rabbit@u1b2"],"feature_flags":[{"desc":"Default bindings are now implicit, instead of being stored in the database","doc_url":"","name":"implicit_default_bindings","provided_by":"rabbit","stability":"stable","state":"enabled"},{"desc":"Maintenance mode status","doc_url":"","name":"maintenance_mode_status","provided_by":"rabbit","stability":"stable","state":"enabled"},{"desc":"Support queues of type `quorum`","doc_url":"https://www.rabbitmq.com/quorum-queues.html","name":"quorum_queue","provided_by":"rabbit","stability":"stable","state":"enabled"},{"desc":"Virtual host metadata (description, tags, etc)","doc_url":"","name":"virtual_host_metadata","provided_by":"rabbit","stability":"stable","state":"enabled"}],"listeners":{"rabbit@u1a2":[{"interface":"[::]","node":"rabbit@u1a2","port":25672,"protocol":"clustering","purpose":"inter-node and CLI tool communication"},{"interface":"[::]","node":"rabbit@u1a2","port":5672,"protocol":"amqp","purpose":"AMQP 0-9-1 and AMQP 1.0"}],"rabbit@u1a3":[{"interface":"[::]","node":"rabbit@u1a3","port":25672,"protocol":"clustering","purpose":"inter-node and CLI tool communication"},{"interface":"[::]","node":"rabbit@u1a3","port":5672,"protocol":"amqp","purpose":"AMQP 0-9-1 and AMQP 1.0"}],"rabbit@u1b2":[{"interface":"[::]","node":"rabbit@u1b2","port":25672,"protocol":"clustering","purpose":"inter-node and CLI tool communication"},{"interface":"[::]","node":"rabbit@u1b2","port":5672,"protocol":"amqp","purpose":"AMQP 0-9-1 and AMQP 1.0"}]},"maintenance_status":{"rabbit@u1a2":"not under maintenance","rabbit@u1a3":"not under maintenance","rabbit@u1b2":"not under maintenance"},"partitions":{},"ram_nodes":[],"running_nodes":["rabbit@u1b2","rabbit@u1a3","rabbit@u1a2"],"versions":{"rabbit@u1a2":{"erlang_version":"23.1.4","rabbitmq_name":"RabbitMQ","rabbitmq_version":"3.8.9"},"rabbit@u1a3":{"erlang_version":"23.1.4","rabbitmq_name":"RabbitMQ","rabbitmq_version":"3.8.9"},"rabbit@u1b2":{"erlang_version":"23.1.4","rabbitmq_name":"RabbitMQ","rabbitmq_version":"3.8.9"}}}
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq/check-version-debian.yml
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq/check-version-debian.yml]: Would this be possible to use here `package_facts` https://docs.ansible.com/ansible/latest/collections/ansible/builtin/package_facts_module.html ? :thinking:

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq/check-version-redhat.yml
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq/check-version-redhat.yml]: Would this be possible to use here `package_facts` https://docs.ansible.com/ansible/latest/collections/ansible/builtin/package_facts_module.html ? :thinking:

**Review by sk4zuzu — CHANGES_REQUESTED** [review]

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml]: Could we use something dedicated to compare versions https://docs.ansible.com/ansible/latest/user_guide/playbooks_tests.html#comparing-versions ? :thinking:

**Review by sk4zuzu — CHANGES_REQUESTED** [review]

**Comment by przemyslavic** [comment]
And what about running `rabbitmq-diagnostics server_version` command to check rabbitmq server version?

`root@2334820499a8:/# rabbitmq-diagnostics server_version`
`Asking node rabbit@2334820499a8 for its RabbitMQ version...`
`3.8.9`

`root@fba4e177a16f:/# rabbitmq-diagnostics server_version`
`Asking node rabbit@fba4e177a16f for its RabbitMQ version...`
`3.7.10`

**Comment by atsikham** [comment]
Additional comment from @przemyslavic:

> What about an upgrade that fails for some reason, but installs new packages before that. The next time the upgrade is started, it will be skipped, because the package version will be up-to-date (and this condition is checked). From what I saw in other components, >= conditions were used just in case the upgrade was interrupted.

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml
Comment (atsikham) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml]: As now we have the version in the correct and same format for all distributions, it's done so - please check.

**Comment by atsikham** [comment]
> And what about running `rabbitmq-diagnostics server_version` command to check rabbitmq server version?
> 
> `root@2334820499a8:/# rabbitmq-diagnostics server_version`
> `Asking node rabbit@2334820499a8 for its RabbitMQ version...`
> `3.8.9`
> 
> `root@fba4e177a16f:/# rabbitmq-diagnostics server_version`
> `Asking node rabbit@fba4e177a16f for its RabbitMQ version...`
> `3.7.10`

Used rabbitmqctl instead with json formatter as @sk4zuzu suggested.

**Review by sk4zuzu — DISMISSED** [review]
LGTM

**Review by przemyslavic — DISMISSED** [review]

**Comment by sk4zuzu** [comment]
@atsikham dont we need changelog entry here ? :thinking:

**Comment by atsikham** [comment]
> @atsikham dont we need changelog entry here ? 🤔

As we had a release and refer to these bugs in the changelog as "known issues", I will update the changelog.

@przemyslavic can we test it somehow before merge?

**Comment by przemyslavic** [comment]
I'll try to test this PR today. Let's not merge it for now.

**Comment by atsikham** [comment]
Tested manually on aws for clustered and not clustered installations.

**Comment by przemyslavic** [comment]
@atsikham still doesn't work for v3.7.10:
```
2021-01-22T12:47:32.7905997Z[38;21m12:47:32 INFO cli.engine.ansible.AnsibleCommand - TASK [upgrade : RabbitMQ | Set facts] ******************************************[0m
2021-01-22T12:47:32.9181086Z[31;21m12:47:32 ERROR cli.engine.ansible.AnsibleCommand - fatal: [ec2-63-35-172-154.eu-west-1.compute.amazonaws.com]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'rabbitmq_version'\n\nThe error appears to be in '/shared/build/06todevawubcanal/ansible/roles/upgrade/tasks/rabbitmq.yml': line 17, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: RabbitMQ | Set facts\n  ^ here\n"}[0m
2021-01-22T12:47:32.9344032Z[31;21m12:47:32 ERROR cli.engine.ansible.AnsibleCommand - fatal: [ec2-18-200-195-163.eu-west-1.compute.amazonaws.com]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'rabbitmq_version'\n\nThe error appears to be in '/shared/build/06todevawubcanal/ansible/roles/upgrade/tasks/rabbitmq.yml': line 17, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: RabbitMQ | Set facts\n  ^ here\n"}
```
No `rabbitmq_version` key found in json output.

**Comment by to-bar** [comment]
How about using lock file approach (upgrade in progress flag file) instead of this condition:
`when: rabbitmq_version_current is version(versions.general, '<=')`

The flag file is used for example for ZooKeeper, IMO is a smart way 🏃🏻

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml:19]: Is it an assumption or a fact? Rabbitmq uses the Raft protocol can we safely assume 0-th node to be a master here? :thinking:

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml:41]: Would be nice to have at least "mode:" here I think? :thinking: Linter will give you a warning without it.

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml
Comment (atsikham) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml:19]: The answer is [here](https://www.rabbitmq.com/upgrade.html#full-stop-upgrades):

```
During an upgrade, the last disc node to go down must be the first node to be brought online. Otherwise the started node will emit an error message and fail to start up. Unlike an ordinary cluster restart, upgrading nodes will not wait for the last disc node to come back online.

While not strictly necessary, it is a good idea to decide ahead of time which disc node will be the upgrader, stop that node last, and start it first. Otherwise changes to the cluster configuration that were made between the upgrader node stopping and the last node stopping will be lost.
```

So it's just a decision to mark the "master"  node as an "upgrader".

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml
Comment (atsikham) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml:41]: Checked that, there is no any linter warnings

```
vscode@3919934ec6c1:/workspaces/epiphany/core/src/epicli$ /home/vscode/.local/bin/ansible-lint data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml
vscode@3919934ec6c1:/workspaces/epiphany/core/src/epicli$ 
```

Comment (atsikham) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml:41]: Modified file permissions

Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/rabbitmq.yml:41]:
```shell
cd data/common/ansible/playbooks/roles/ && /home/vscode/.local/bin/ansible-lint upgrade
```
works much better :+1::innocent:

**Review by sk4zuzu — DISMISSED** [review]
:heart:

**Review by sk4zuzu — APPROVED** [review]

**Review by przemyslavic — APPROVED** [review]

**Comment by przemyslavic** [comment]
@atsikham tested and approved :)

**Pull request merged**
